File: assignmentHotelRoom.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

unordered_map<ll, ll> p;

ll find(ll x) {
    if (!p[x]) {
        return x;
    }
    return p[x] = find(p[x]);
}

void merge(ll x, ll y) {
    x = find(x);
    y = find(y);
    if (x == y) return;
    if (x > y) swap(x, y);
    p[x] = y;
}

vector<ll> solution(ll k, vector<ll> room_number) {
    vector<ll> answer;
    for (auto& r : room_number) {
        if (!p[r]) {
            answer.push_back(r);
            merge(r, r + 1);
        } else {
            ll tf = find(r);
            answer.push_back(tf);
            merge(tf, tf + 1);
        }
    }
    return answer;
}

File: assignmentHotelRoom.js
let parent = new Map();

function find(x) {
    if (parent.get(x) === undefined) {
        return x;
    }
    let tmp = find(parent.get(x));
    parent.set(x, tmp);
    return tmp;
}

function merge(x, y) {
    x = find(x);
    y = find(y);
    if (x === y) {
        return;
    }
    if (x > y) {
        [x, y] = [y, x];
    }
    parent.set(x, y);
}

function solution(k, room) {
    let ans = [];
    for (let i = 0; i < room.length; i++) {
        let tmp = find(room[i]);
        if (tmp) {
            merge(tmp, tmp + 1);
            ans.push(tmp);
        } else {
            merge(room[i], room[i] + 1);
            ans.push(room[i]);
        }
    }
    return ans;
}
